### PR TITLE
[shopsys] made parameters overridable by ENV variables

### PR DESCRIPTION
**Pull request by s3tezsky**
| Q             | A
| ------------- | ---
|Description, reason for the PR| We want to be able change parameters via ENV variables to easier change setting of application in any environment. This PR allows developers to override all parameters in `parameters.yaml` by ENV variables. This PR also unifies usage of parameters in different environment (dev, test, prod).
|New feature| Yes 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

The original PR (#1615) was braking backward compatibility.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -186,12 +185,14 @@
     },
     "scripts": {
         "post-install-cmd": [
-            "Incenteev\\ParameterHandler\\ScriptHandler::buildParameters",
+            "cp -n project-base/config/parameters.yaml.dist project-base/config/parameters.yaml",
+            "cp -n project-base/config/parameters_test.yaml.dist project-base/config/parameters_test.yaml",
```
Comment (grossmannmartin) [inline composer.json:189]: `cp -n ` returns 1 on mac. It's impossible for me to install application if parameters.yaml (parameters_test.yaml) exists. 

it could work with 
```
cp -n project-base/config/parameters.yaml.dist project-base/config/parameters.yaml || true
```
or 
```
cp -n project-base/config/parameters.yaml.dist project-base/config/parameters.yaml || echo '[skiped] parameters.yaml exists'
```

File: project-base/tests/App/Acceptance/acceptance.suite.yml [hunk 2]
```diff
@@ -32,7 +32,7 @@ modules:
             populate: true
             cleanup: true
         Tests\App\Test\Codeception\Module\StrictWebDriver:
-            host: "%selenium_server_host%"
+            host: '%SELENIUM_SERVER_HOST%'
```
Comment (grossmannmartin) [inline project-base/tests/App/Acceptance/acceptance.suite.yml:35]: this parameter is lowercased in `parameters_test.yaml.dist`. Is this change intentional?

File: project-base/composer.json [hunk 3]
```diff
@@ -49,7 +49,6 @@
         "fzaninotto/faker": "^1.7.1",
         "helios-ag/fm-elfinder-bundle": "^10.0.4",
         "heureka/overeno-zakazniky": "^2.0.6",
-        "incenteev/composer-parameter-handler": "^2.1.3",
```
Comment (grossmannmartin) [inline project-base/composer.json:52]: This dependency should be removed even from symfony.lock file (both monorepo/project-base)

File: project-base/.env [hunk 4]
```diff
@@ -14,5 +14,23 @@
...
+
+MAILER_TRANSPORT=smtp
+MAILER_HOST=smtp-server
+MAILER_USER=
+MAILER_PASSWORD=
+MAILER_DELIVERY_WHITELIST=/@shopsys\.com$/
+MAILER_DISABLE_DELIVERY=0
+MAILER_MASTER_EMAIL_ADDRESS=no-reply@shopsys.com
+
+SECRET=ThisTokenIsNotSoSecretChangeIt
```
Comment (grossmannmartin) [inline project-base/.env:35]: Symfony by default uses `APP_SECRET`. Is a reason to not follow with the same name?

File: upgrade/UPGRADE-v9.1.0-dev.md [hunk 5]
```diff
@@ -141,3 +141,6 @@ There you can find links to upgrade notes for other versions too.
 
 - made tests to be domain independent ([#2051](https://github.com/shopsys/shopsys/pull/2051))
     - see #project-base-diff to update your project
+
+- made parameters overridable by ENV variables ([#2055](https://github.com/shopsys/shopsys/pull/2055))
+    - see #project-base-diff to update your project
```
Comment (grossmannmartin) [inline upgrade/UPGRADE-v9.1.0-dev.md:146]: I think a little info about where set the values instead of parameters.yaml file will be nice. Maybe a link to the documentation.

**Review by grossmannmartin — CHANGES_REQUESTED** [review]
Hi, thanks for the change. I like the backward compatibility kept 👍 

Can you please file an issue (or solve right now, if you feel like it) to address the application environment (DEV, TEST, PROD) and debug mode set with the environment variables? It was part of #1615 and I wouldn't want to forget about it.

File: project-base/.env [hunk 6]
```diff
@@ -14,5 +14,23 @@
...
 ELASTIC_SEARCH_INDEX_PREFIX=''
+REDIS_HOST=redis
+REDIS_PREFIX=''
+
+MAILER_TRANSPORT=smtp
+MAILER_HOST=smtp-server
+MAILER_USER=
+MAILER_PASSWORD=
+MAILER_DELIVERY_WHITELIST=/@shopsys\.com$/
+MAILER_DISABLE_DELIVERY=0
```
Comment (grossmannmartin) [inline project-base/.env:32]: I can be tempted to set `false` here and I think it won't work. Not sure whether it is a problem

File: project-base/config/bootstrap.php [hunk 7]
```diff
@@ -19,7 +19,7 @@
     throw new RuntimeException('Please run "composer require symfony/dotenv" to load the ".env" files configuring the application.');
 } else {
     $path = dirname(__DIR__).'/.env';
-    $dotenv = new Dotenv(false);
+    $dotenv = new Dotenv(true);
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply project-base/config/bootstrap.php:22]: It is. It is not really easy to find this information, but without it the `getenv` will return false and only way how to get it is through globals.

File: project-base/tests/App/Acceptance/acceptance.suite.yml (hunk 2 above)
Comment (s3tezsky) [reply project-base/tests/App/Acceptance/acceptance.suite.yml:35]: Yes it is intensional. This variable has value from ENV. It is weird but that is how the codeception works.

Anyway I have removed original parameter `selenium_server_host` because it is not used anymore.

File: project-base/build/codeception.yml [hunk 8]
```diff
@@ -6,6 +6,8 @@ paths:
     data: ../tests/App/Acceptance/data
     support: ../tests/App/Test/Codeception
 params:
+    - ../.env
+    - ../.env.test
     - ../config/parameters_test.yaml
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply project-base/build/codeception.yml:11]: It is. In that file there are mapped ENVs to current parameter. I have added deprecation comment above.

File: project-base/.env (hunk 4 above)
Comment (s3tezsky) [reply project-base/.env:35]: No reason. I kept original name. Anyway renaming it to `APP_SECRET` is really good point 👍

File: .ci/build_kubernetes.sh [hunk 9]
```diff
@@ -16,9 +16,6 @@ cp project-base/config/parameters.yaml.dist project-base/config/parameters.yaml
 yq write --inplace project-base/config/domains_urls.yaml domains_urls[0].url http://${FIRST_DOMAIN_HOSTNAME}:${NGINX_INGRESS_CONTROLLER_HOST_PORT}
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply .ci/build_kubernetes.sh:16]: Not sure why is it pointed here but ok.

Files `parameters(_test).yaml` cannot be removed yet. There is mapping from ENVs to parameters which are later used for service configuration and so on.

File: composer.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (s3tezsky) [reply composer.json]: fixed

File: docs/cookbook/dumping-and-importing-the-database.md [hunk 10]
```diff
@@ -30,7 +30,7 @@ Then you can import the dump:
 psql --quiet --username=database_user --host=database_host target_database_name < dump.sql
 ```
 
-Replace `database_user`, `database_host` and `target_database_name` with the correct values (from your `config/parameters.yaml`).
+Replace `database_user`, `database_host` and `target_database_name` with the correct values (from your `.env.locaal`).
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply docs/cookbook/dumping-and-importing-the-database.md]: fixed

File: docs/introduction/running-acceptance-tests.md [hunk 11]
```diff
@@ -10,7 +10,7 @@ docker exec -it shopsys-framework-php-fpm bash
 
 !!! note
     For `selenium-server` to be able to connect to you `webserver` container and access your application, all domains should have URL set to `http://webserver:8000`.
-    This is done via parameter `%overwrite_domain_url%` defined in `config/parameters_test.yaml`.
+    This is done via parameter `%overwrite_domain_url%` defined in `.env.test`.
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply docs/introduction/running-acceptance-tests.md]: fixed

File: project-base/config/parameters.yaml.dist [hunk 12]
```diff
@@ -1,22 +1,20 @@
+# @deprecated parameters will be set by ENV variables only in next major version
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply project-base/config/parameters.yaml.dist]: fixed

File: project-base/composer.json [hunk 13]
```diff
@@ -126,12 +125,14 @@
     },
     "scripts": {
         "post-install-cmd": [
-            "Incenteev\\ParameterHandler\\ScriptHandler::buildParameters",
+            "cp -n config/parameters.yaml.dist config/parameters.yaml",
+            "cp -n config/parameters_test.yaml.dist config/parameters_test.yaml",
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply project-base/composer.json]: fixed

File: composer.json (hunk 1 above)
Comment (s3tezsky) [reply composer.json]: fixed

File: project-base/composer.json (hunk 3 above)
Comment (s3tezsky) [reply project-base/composer.json:52]: fixed

File: project-base/.env (hunk 4 above)
Comment (s3tezsky) [reply project-base/.env]: fixed

File: packages/framework/src/Twig/MailerSettingExtension.php [hunk 14]
```diff
@@ -41,9 +41,28 @@ class MailerSettingExtension extends AbstractExtension
...
+            $this->isDeliveryDisabled = $this->container->getParameter('mailer_disable_delivery');
+        }
+
+        $envMailerMasterEmailAddress = getenv('MAILER_MASTER_EMAIL_ADDRESS');
+        if ($envMailerMasterEmailAddress !== false) {
+            $this->mailerMasterEmailAddress = $envMailerMasterEmailAddress;
+        } else {
+            $this->mailerMasterEmailAddress = $this->container->getParameter('mailer_master_email_address');
+        }
+
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply packages/framework/src/Twig/MailerSettingExtension.php]: Yeah, nice catch! Fixed.

**Comment by s3tezsky** [comment]
> Can you please file an issue (or solve right now, if you feel like it) to address the application environment (DEV, TEST, PROD) and debug mode set with the environment variables? It was part of #1615 and I wouldn't want to forget about it.

At first I wantet to have it implemented here, then I thought about an effort and potantial risks so I have created an issue here https://github.com/shopsys/shopsys/issues/2065.

File: composer.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (s3tezsky) [reply composer.json]: fixed

File: upgrade/UPGRADE-v9.1.0-dev.md (hunk 5 above)
Comment (s3tezsky) [reply upgrade/UPGRADE-v9.1.0-dev.md:147]: fixed

File: project-base/.env (hunk 6 above)
Comment (s3tezsky) [reply project-base/.env:32]: I am not really sure if there is something what we can do about it. The ENVs are string only.

File: project-base/build/codeception.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply project-base/build/codeception.yml:12]: It seems the parameters are loaded from .env files directly (I am able to run acceptance tests without parameters_test.yaml file loaded).

File: project-base/.env (hunk 6 above)
Comment (grossmannmartin) [reply project-base/.env:32]: Symfony uses env processor (:bool) which accepts `no`, `false`, etc., so that's why I can be tempted to do this here.

But I'm not sure if it's something we can do here. Maybe a comment or documentation

File: docs/introduction/running-acceptance-tests.md [hunk 15]
```diff
@@ -10,7 +10,7 @@ docker exec -it shopsys-framework-php-fpm bash
 
 !!! note
     For `selenium-server` to be able to connect to you `webserver` container and access your application, all domains should have URL set to `http://webserver:8000`.
-    This is done via parameter `%overwrite_domain_url%` defined in `config/parameters_test.yaml`.
+    This is done via parameter `%overwrite_domain_url%` defined in `.env.test` or `.env.test.local`.
```
Comment (grossmannmartin) [inline docs/introduction/running-acceptance-tests.md:13]:
```suggestion
    This is done via parameter `OVERWRITE_DOMAIN_URL` defined in `.env.test` or `.env.test.local`.
```

File: project-base/config/parameters.yaml.dist [hunk 16]
```diff
@@ -1,22 +1,17 @@
...
+    database_password: '%env(DATABASE_PASSWORD)%'
+    elasticsearch_host: '%env(ELASTICSEARCH_HOST)%'
+    redis_host: '%env(REDIS_HOST)%'
+    mailer_transport: '%env(MAILER_TRANSPORT)%'
+    mailer_host: '%env(MAILER_HOST)%'
+    mailer_user: '%env(MAILER_USER)%'
+    mailer_password: '%env(MAILER_PASSWORD)%'
     mailer_disable_delivery: false
     mailer_master_email_address: no-reply@shopsys.com
     mailer_delivery_whitelist: [/@shopsys\.com$/]
```
[the remark this answers is not in the document]
Comment (henzigo) [reply project-base/config/parameters.yaml.dist:16]: Oh, now I see that those parameters can be set by ENV. But I think that this is a little bit confusing

File: project-base/.env [hunk 17]
```diff
@@ -14,5 +14,23 @@
 # https://symfony.com/doc/current/best_practices/configuration.html#infrastructure-related-configuration
 
 IGNORE_DEFAULT_ADMIN_PASSWORD_CHECK='0'
-REDIS_PREFIX=''
+DATABASE_HOST=postgres
+DATABASE_PORT=5432
+DATABASE_NAME=shopsys
+DATABASE_USER=root
+DATABASE_PASSWORD=root
```
Comment (henzigo) [inline project-base/.env:21]: Why are we do not connection string instead of those parameters? 

We would really appreciate this because our projects have a lot of ENV variables and we are lost with it. Moreover, Symfony is propagating connection strings and there is not any benefit for do not use connection string

Please consider using it ❤️

File: project-base/.env [hunk 18]
```diff
@@ -14,5 +14,23 @@
...
+ELASTICSEARCH_HOST='elasticsearch:9200'
 ELASTIC_SEARCH_INDEX_PREFIX=''
+REDIS_HOST=redis
+REDIS_PREFIX=''
+
+MAILER_TRANSPORT=smtp
+MAILER_HOST=smtp-server
+MAILER_USER=
+MAILER_PASSWORD=
```
Comment (henzigo) [inline project-base/.env:30]: Same as above

File: docs/installation/application-configuration.md [hunk 19]
```diff
@@ -1,71 +1,49 @@
...
+| `ELASTIC_SEARCH_INDEX_PREFIX`          | `''`                                 | separates more projects that use the same elasticsearch service                      |
+| `IGNORE_DEFAULT_ADMIN_PASSWORD_CHECK`  | `'0'`                                | set to `true` if you want to allow administrators to log in with default credentials |
+| `OVERWRITE_DOMAIN_URL`                 | `'http://webserver:8080'`            | overwrites URL of all domains for acceptance testing (set to `~` to disable)         |
+| `SELENIUM_SERVER_HOST`                 | `'selenium-server'`                  | with native installation the selenium server is on `localhost`                       |
+| `SHOPSYS_CONTENT_DIR_NAME`             | `'content-test'`                     | web/content-test/ directory is used instead of web/content/ during the tests         |
+| `TRUSTED_PROXIES`                      | `'127.0.0.1'`                        | proxies that are trusted to pass traffic, used mainly for production (set as text separated by comma for multiple values) |
+
```
Comment (henzigo) [inline docs/installation/application-configuration.md:46]: Maybe there would be some info about dumping ENV variables with `php bin/console debug:container --env-vars` command

File: project-base/.env.test [hunk 20]
```diff
@@ -1 +1,14 @@
 # define your env variables for the test environment here
+
+DATABASE_HOST=postgres
+DATABASE_PORT=5432
+DATABASE_NAME=shopsys-test
+DATABASE_USER=root
+DATABASE_PASSWORD=root
```
Comment (henzigo) [inline project-base/.env.test:7]: You do not need to have duplicated variables here. Even in `TEST` environment, `.env` file is loaded first, so you do not need to override variables that are the same

File: project-base/.env.test [hunk 21]
```diff
@@ -1 +1,14 @@
...
+SELENIUM_SERVER_HOST=selenium-server
+OVERWRITE_DOMAIN_URL=http://webserver:8080
+MAILER_TRANSPORT=smtp
+MAILER_HOST=smtp-server
+MAILER_USER=
+MAILER_PASSWORD=
+SHOPSYS_CONTENT_DIR_NAME=content-test
```
Comment (henzigo) [inline project-base/.env.test:14]: I'm missing here variables with default prefixes for Elastic and Redis.

**Review by henzigo — COMMENTED** [review]
Hello, 
I just added some points from me, so please look at them.

File: project-base/composer.json [hunk 22]
```diff
@@ -126,12 +125,14 @@
     },
     "scripts": {
         "post-install-cmd": [
-            "Incenteev\\ParameterHandler\\ScriptHandler::buildParameters",
+            "cp -n config/parameters.yaml.dist config/parameters.yaml || echo '[skiped] parameters.yaml exists'",
```
[the remark this answers is not in the document]
Comment (s3tezsky) [reply project-base/composer.json:128]: I have discribed it in https://github.com/shopsys/shopsys/pull/2055#discussion_r501597793

You should not really care about parameter files but they are necessary for keeping backward compatibility.

File: project-base/config/parameters.yaml.dist (hunk 16 above)
[the remark this answers is not in the document]
Comment (s3tezsky) [reply project-base/config/parameters.yaml.dist:16]: Sadly there is not better option to achieve that, because of swift mailer. I am sad as a developer of SSFW by doing so. Anyway as implementator I do not need to care about it becase it works as it is described in our doc.

File: project-base/.env (hunk 17 above)
Comment (s3tezsky) [reply project-base/.env:21]: We have considered it and decide to not implement it now. We do not want to change the behavior except passing parameters.

I personally like configuring connection by setting each parameter separate becouse I cannot remember the schema of connection string so I have to google it every time.

File: project-base/.env (hunk 18 above)
Comment (s3tezsky) [reply project-base/.env:30]: Same as above :)

File: docs/installation/application-configuration.md (hunk 19 above)
Comment (s3tezsky) [reply docs/installation/application-configuration.md:46]: That is right, even it is in symfony docs we could mention it here.

File: project-base/.env.test (hunk 20 above)
Comment (s3tezsky) [reply project-base/.env.test:7]: I am confused how should the .env file should look like. In this comment you mind redundant ENVs but in comment bellow you want to introduce more reduntant ENVs. 🤔

File: project-base/.env.test (hunk 21 above)
Comment (s3tezsky) [reply project-base/.env.test:14]: I am resolving this one. Pleaso continue discussion above https://github.com/shopsys/shopsys/pull/2055#discussion_r502387455

File: project-base/.env.test (hunk 20 above)
Comment (henzigo) [reply project-base/.env.test:7]: Oh, I see. I want to remove redundant parameters, but introduce prefixes that should be different by default (probably with `test-` prefix).

Comment (s3tezsky) [reply project-base/.env.test:7]: I am for removing the reduntant ENVs. It means also not introducing test prefixes as we do not have them configured. Maybe we should get it done in following PR. Are you OK with that?

File: project-base/.env (hunk 6 above)
Comment (s3tezsky) [reply project-base/.env:32]: As this is problem only when ENV is used via `getenv()` and this is only usage I even more hacked our hack.

IMHO it is ok to do that without any regrets as it is for unified usages of ENVs and to prevent confusion or misconfiguration.

So what is it?

Simply retype the returned value from `getenv()` and check if it is not `string(5) "false"`.

```
$envMailerDisableDelivery = getenv('MAILER_DISABLE_DELIVERY');
        if ($envMailerDisableDelivery !== false) {
            $castedEnvMailerDisableDelivery = (bool)$envMailerDisableDelivery === true && $envMailerDisableDelivery !== 'false';
            $container->setParameter('mailer_disable_delivery', $castedEnvMailerDisableDelivery);
        }
```

Are you OK with that? It will do much better than some tiny note in doc. And even better to see happy developers.

File: docs/installation/application-configuration.md (hunk 19 above)
Comment (s3tezsky) [reply docs/installation/application-configuration.md:49]: Added in 8d21964ae8c0c934939fefb5ff934da24cb351e5

File: project-base/.env (hunk 17 above)
Comment (s3tezsky) [reply project-base/.env:21]: Redundats were removed in 2f196da2d7881ea636a8c0972216fb5f366dea27

File: project-base/.env.test (hunk 20 above)
Comment (s3tezsky) [reply project-base/.env.test]: Redundats were removed in 2f196da2d7881ea636a8c0972216fb5f366dea27

File: project-base/build/codeception.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (s3tezsky) [reply project-base/build/codeception.yml]: Fixed in ba394bee4433298d9db2e2be78ad9ae820f71c78

File: project-base/.env (hunk 6 above)
Comment (s3tezsky) [reply project-base/.env:32]: Implemented in 71e13a9cdec3d6e81dccd8332766181005ee1f1a

Comment (grossmannmartin) [reply project-base/.env:32]: That's cool. I did take a look at the implementation of the env processor in Symfony (`\Symfony\Component\DependencyInjection\EnvVarProcessor`) and for bool they're using 
```
(bool) (filter_var($env, \FILTER_VALIDATE_BOOLEAN) ?: filter_var($env, \FILTER_VALIDATE_INT) ?: filter_var($env, \FILTER_VALIDATE_FLOAT));
```
does it make sense to use this for `$castedEnvMailerDisableDelivery`? It will be unified with the way, how Symfony handles boolean environment variable values (for example 'no' is also false)

File: .ci/build_kubernetes.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply .ci/build_kubernetes.sh:16]: I see, in that case, should it be deprecated? 

I see a deprecation as "not used, should not be used in a new code, but kept to make your old code work".
Here it's still a crucial file for the application to work

File: docs/introduction/running-acceptance-tests.md (hunk 15 above)
Comment (grossmannmartin) [reply docs/introduction/running-acceptance-tests.md:13]: I don't see this applied, is it intentional or by mistake?

File: project-base/composer.json (hunk 22 above)
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply project-base/composer.json:128]: I feel like deprecation of this file cause unnecessary confusion,  so I suggest to change the message in those files to describe what is hapening

File: project-base/.env (hunk 17 above)
Comment (henzigo) [reply project-base/.env:21]: I would like to have this feature in next major release. Is this possible? There can be comment about "How to use connection string". 

Can you create an issue for that? Or can we discuss it, because for us this little change has really big advantage

Comment (s3tezsky) [reply project-base/.env:21]: Ok, I have created an issue https://github.com/shopsys/shopsys/issues/2072.

I am resolving this.

File: project-base/composer.json (hunk 22 above)
[the remark this answers is not in the document]
Comment (s3tezsky) [reply project-base/composer.json:128]: changed the deprecation message in c70716c698a93c6d52d757a99f204f0f8b03b162

File: docs/introduction/running-acceptance-tests.md (hunk 15 above)
Comment (s3tezsky) [reply docs/introduction/running-acceptance-tests.md]: Yah, my bad. Fixed 4dbd0653a093429922a5d1d0283aad87d3ba51ae

File: project-base/.env (hunk 6 above)
Comment (s3tezsky) [reply project-base/.env:32]: Nice! Fixed in 4ade8e8da9fae85d4242ed6c764f4ff33f9724b3

**Review by grossmannmartin — APPROVED** [review]

**Comment by grossmannmartin** [comment]
Hey, I have one point from testing, can you please validate it?

To propagate the change of environment variables `MAILER_DISABLE_DELIVERY`, `MAILER_DELIVERY_WHITELIST`, `MAILER_MASTER_EMAIL_ADDRESS` I need to clean the cache
-

File: docs/installation/application-configuration.md [hunk 23]
```diff
@@ -22,6 +22,9 @@ You may want to set some settings in a different way (such as production, test,
 
 ### Application
 
+!!! note
+    To propagate the change of `MAILER_DELIVERY_WHITELIST`, `MAILER_DISABLE_DELIVERY` and `MAILER_MASTER_EMAIL_ADDRESS` cleaning cache must be cleaned.
```
Comment (grossmannmartin) [inline docs/installation/application-configuration.md:26]:
```suggestion
    To propagate the change of `MAILER_DELIVERY_WHITELIST`, `MAILER_DISABLE_DELIVERY` and `MAILER_MASTER_EMAIL_ADDRESS` cache must be cleaned. 
```

**Review by grossmannmartin — APPROVED** [review]

**Review by sslt — APPROVED** [review]
